doc/advanced-testing-mini-course/src/Lec05/EventQueue.hs

```diff
@@ -33,11 +45,22 @@ fakeEventQueue a0 _clock = do
     enqueue :: IORef Agenda -> Event -> IO ()
     enqueue agenda event = modifyIORef' agenda (push (getEventTime event, event))
 
-    dequeue :: IORef Agenda -> IO Event
-    dequeue agenda = do
+    dequeue :: IORef Agenda -> DequeueTimeout -> IO Event
+    dequeue agenda dequeueTimeout = do
       a <- readIORef agenda
       case pop a of
-        Nothing -> return (CommandEventE Exit)
-        Just ((_time, event), a') -> do
-          writeIORef agenda a'
-          return event
+        Nothing -> case dequeueTimeout of
+          NoTimeout -> return (CommandEventE Exit)
+          Timeout _micros retry -> retry
```

Comment:
We need some way of stopping the simulation, if I keep registering timers the simulation will never end.

Reply:
We should probably always add a `Exit` event in the agenda at the time we want the simulation to end. Let's fix this in a separate PR.